### ISSUE TITLE
Builtin visualization, slows Perspective Transformer to 40 seconds per step

### ISSUE THREAD

**Issue by mees** [tensorflow/models#2221]
### System information
- **What is the top-level directory of the model you are using**: Ptn directory
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 14.04
- **TensorFlow installed from (source or binary)**: Source 
- **TensorFlow version (use command below)**: ('v1.3.0-rc1-27-g2784b1c', '1.3.0-rc2')
- **Bazel version (if compiling from source)**: 0.5.2
- **CUDA/cuDNN version**: 8.0/5.1
- **GPU model and memory**: Titan X
- **Exact command to reproduce**: CUDA_VISIBLE_DEVICES=0 bazel-bin/train_ptn   '--checkpoint_dir=/home/meeso/models/ptn/my_models/no_finetune' '--inp_dir=/tmp/shapenet_tf/' '

### Describe the problem
write_disk_op is computed on each step, generating the 3D visualization for tensorboard on each training step. The visualization runs on CPU and takes a long time to compute. There should be an option to execute the visualization script only every n iterations (best passed as a flag), something like if global_step % .FLAGS.visualize_every==0, then execute visualization. I tried fixing it with tf.cond, but was unsuccessful. With visualization, i.e. the code in the official repo right now, I get 39.396 sec/step vs (0.220 sec/step without the visualization) and 0.02 steps/ sec in tensorboard.

**Comment by mees** [comment]
@xcyan, @arkanath

**Comment by xcyan** [comment]
We didn't observe the phenomenon on our side. Basically, the performance depends on the CPU you are using. If you find the bottleneck is CPU computation, please move the grid computation inside the "if" and returns an empty grid if needed.

**Comment by mees** [comment]
Well the server has 2x Intel(R) Xeon(R) CPU E5-2620 v4 @ 2.10GHz, so I don't think that the problem is that the CPU is too slow. On the contrary, if with such high performance CPU the visualization code needs so long to run, there might be a bug in the visualization code. I can help you debug it if you want, or you could write a workaround such that the visualization is only executed every x steps. As a workaround I have created a flag to either enable or disable the visualization script, but this is a suboptimal temporary solution.

**Comment by xcyan** [comment]
It seems that data loading is the bottleneck. For visualization, you need another 4 readers to load batches from valid set. We didn't observe such difficulty in google machines. But if you want to speed up the runtime on your side, you can try to reduce the number of samples during validation.

**Comment by mees** [comment]
@xcyan I am not sure that the data loader is the problem, because when I disable the visualization, I am still loading the valid data and the loss is computed and shown in tensorboard. As you can see in my code, the data loading with the 4 threads happens in model.get_inputs(), which is executed still even when I disable the visualization. Thus it appears as the bottleneck is the visualization code itself right?

`      if FLAGS.task == 0:
        params = FLAGS
        params.batch_size = params.num_views
        params.step_size = 1
        model.set_params(params)
        val_data = model.get_inputs(
            params.inp_dir,
            params.dataset_name,
            'val',
            params.batch_size,
            params.image_size,
            params.vox_size,
            is_training=False)
        val_inputs = model.preprocess(val_data, params.step_size)
        # Note: don't compute loss here
        reused_model_fn = model.get_model_fn(is_training=False, reuse=True)
        val_outputs = reused_model_fn(val_inputs)
      if FLAGS.visualize:
        with tf.device(tf.DeviceSpec(device_type='CPU')):
          vis_input_images = val_inputs['images_1'] * 255.0
          vis_gt_projs = (val_outputs['masks_1'] * (-1) + 1) * 255.0
          vis_pred_projs = (val_outputs['projs_1'] * (-1) + 1) * 255.0

          vis_gt_projs = tf.concat([vis_gt_projs] * 3, axis=3)
          vis_pred_projs = tf.concat([vis_pred_projs] * 3, axis=3)
          # rescale
          new_size = [FLAGS.image_size] * 2
          vis_gt_projs = tf.image.resize_nearest_neighbor(
              vis_gt_projs, new_size)
          vis_pred_projs = tf.image.resize_nearest_neighbor(
              vis_pred_projs, new_size)
          # flip
          # vis_gt_projs = utils.image_flipud(vis_gt_projs)
          # vis_pred_projs = utils.image_flipud(vis_pred_projs)
          # vis_gt_projs is of shape [batch, height, width, channels]
          write_disk_op = model.write_disk_grid(
              global_step=global_step,
              log_dir=save_image_dir,
              input_images=vis_input_images,
              gt_projs=vis_gt_projs,
              pred_projs=vis_pred_projs,
              input_voxels=val_inputs['voxels'],
              output_voxels=val_outputs['voxels_1'])
	with tf.control_dependencies([write_disk_op]):
	  train_op = tf.identity(train_op)
`

**Comment by mees** [comment]
I attach a trace with the visualization on
[tf_trace-10.json.zip](https://github.com/tensorflow/models/files/1227976/tf_trace-10.json.zip)

**Comment by xcyan** [comment]
See my comment above (with code attached).

**Comment by mees** [comment]
Yes that helped @xcyan. I found that I had to store the previous grid temporally to show it in tensorboard while the new one is computed, to avoid showing a black image in between. I also had to reduce the save_summaries_secs parameter since the computation got much faster.

**Comment by skye** [comment]
@mees is this issue resolved?

**Comment by mees** [comment]
I guess, you would need a PR to fix it for everybody.

**Comment by skye** [comment]
@xycan do you think it makes sense to fix this generally? If not please close the issue.

**Comment by xcyan** [comment]
Please close this issue. We will add a note to the README. Thanks!

**Comment by xcyan** [comment]
@mees @arkanath @skye I tried to move the tf.summary inside tf.cond but ran into this issue:
https://github.com/tensorflow/tensorflow/issues/10059

I don't have a better solution for now.
